### PR TITLE
feat: zerion market data service for custom assets

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds Zerion market data service for custom assets.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Relates to https://github.com/shapeshift/web/issues/7245

## Risk

Small

> What protocols, transaction types or contract interactions might be affected by this PR?

None.

## Testing

Add a custom asset that we can expect to have market data for, and confirm that the market data exists in both the trade flow (immediately upon adding the asset) and on the asset page (see video below).

For engineering, the easiest way to do this is to just remove an asset from the generated asset list and then add it as a custom asset. It is also useful to move `new ZerionMarketService(),` in `src/lib/market-service/market-service-manager.ts` to the top once all assets are loaded so that it is indeed Zerion that does the fetching (Portals is also eager to help!).

For ops this is tricky, as most (effectively all?) new assets don't have market data available.

### Engineering

☝️

### Operations

☝️

Confirm that adding an asset with no market data still behaves as it currently does.

## Screenshots (if applicable)

https://github.com/user-attachments/assets/eb51b65e-e965-4ea6-a9da-f18b0b31eed4

### CODE REVIEW

File: src/lib/market-service/portals/portals.ts [hunk 1]
```diff
@@ -33,53 +34,7 @@ const calculatePercentChange = (openPrice: string, closePrice: string): number =
...
-        pendingResolves.push(resolve)
-      })
-    }
-  }
-
-  const clear = () => clearInterval(intervalId)
-
-  return { throttle, clear }
-})({
+const { throttle, clear } = createThrottle({
```
Comment (0xApotheosis) [inline src/lib/market-service/portals/portals.ts:37]: Moved out into a util as we'll reuse this when we add the `findAll` method to the `ZerionMarketService`.

File: src/lib/market-service/utils.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+export const createThrottle = ({
```
Comment (0xApotheosis) [inline src/lib/market-service/utils.ts:1]: Just moved out of the Portals market service.

File: src/lib/market-service/zerion/zerion.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  getZerionTokenMarketData = async (assetId: AssetId): Promise<ZerionMarketData | undefined> => {
+    const { assetReference } = fromAssetId(assetId)
+    const url = `${this.baseUrl}/fungibles/${assetReference}`
+    const { data: res } = await axios.get<ZerionFungibles>(url)
+    return res.data.attributes.market_data
+  }
+
+  async findAll(_args?: FindAllMarketArgs): Promise<MarketCapResult> {
+    // TODO - unimplemented
```
Comment (0xApotheosis) [inline src/lib/market-service/zerion/zerion.ts:31]: For a follow-up PR, less important than getting custom asset market data ready before custom token marketing.

File: src/lib/market-service/zerion/zerion.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    return await Promise.resolve({})
+  }
+
+  async findPriceHistoryByAssetId({
+    assetId: _assetId,
+    timeframe: _timeframe,
+  }: PriceHistoryArgs): Promise<HistoryData[]> {
+    // TODO - unimplemented
```
Comment (0xApotheosis) [inline src/lib/market-service/zerion/zerion.ts:39]: For a follow-up PR, less important than getting custom asset market data ready before custom token marketing.

**Review by gomesalexandre — COMMENTED** [review]
@0xApotheosis looks like this guy is missing a last commit re: misising type imports and util, but I ran this with the diff below as a sanity check (note prioritization of Zerion) and looks to be the goer after missing bits go in! More redundancy 🎉 

https://jam.dev/c/1c0181c2-face-4dbd-afac-7e795b81dd09

On a side-note, is there any asset this can be tested with where market-data is missing on Portals, but present in Zerion, so we can test with production criteria rather than monkey-patching the order?


```diff
diff --git a/src/components/TradeAssetSearch/hooks/useGetCustomTokenPriceQuery.tsx b/src/components/TradeAssetSearch/hooks/useGetCustomTokenPriceQuery.tsx
index 1e0c6f91b2..bd09c54ff5 100644
--- a/src/components/TradeAssetSearch/hooks/useGetCustomTokenPriceQuery.tsx
+++ b/src/components/TradeAssetSearch/hooks/useGetCustomTokenPriceQuery.tsx
@@ -16,7 +16,7 @@ type UseGetCustomTokenPricesQueryReturn = {
 
 const ZERION_BASE_URL = 'https://zerion.shapeshift.com/'
 
-type ZerionMarketData = {
+export type ZerionMarketData = {
   price: number
   total_supply: number
   circulating_supply: number
@@ -30,7 +30,7 @@ type ZerionMarketData = {
   }
 }
 
-type ZerionFungibles = {
+export type ZerionFungibles = {
   data: {
     attributes: {
       market_data: ZerionMarketData
diff --git a/src/lib/asset-service/service/generatedAssetData.json b/src/lib/asset-service/service/generatedAssetData.json
index 94ee0e9a90..bc2180d93a 100644
--- a/src/lib/asset-service/service/generatedAssetData.json
+++ b/src/lib/asset-service/service/generatedAssetData.json
@@ -78652,19 +78652,6 @@
     "explorerTxLink": "https://etherscan.io/tx/",
     "relatedAssetKey": null
   },
-  "eip155:1/erc20:0xae7ab96520de3a18e5e111b5eaab095312d7fe84": {
-    "assetId": "eip155:1/erc20:0xae7ab96520de3a18e5e111b5eaab095312d7fe84",
-    "chainId": "eip155:1",
-    "name": "Lido Staked Ether",
-    "precision": 18,
-    "color": "#04A4FC",
-    "icon": "https://rawcdn.githack.com/trustwallet/assets/master/blockchains/ethereum/assets/0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84/logo.png",
-    "symbol": "STETH",
-    "explorer": "https://etherscan.io",
-    "explorerAddressLink": "https://etherscan.io/address/",
-    "explorerTxLink": "https://etherscan.io/tx/",
-    "relatedAssetKey": null
-  },
   "eip155:1/erc20:0xae8535c23afedda9304b03c68a3563b75fc8f92b": {
     "explorer": "https://etherscan.io",
     "explorerAddressLink": "https://etherscan.io/address/",
@@ -240542,4 +240529,4 @@
     "explorerTxLink": "https://basescan.org/tx/",
     "relatedAssetKey": "eip155:1/slip44:60"
   }
-}
\ No newline at end of file
+}
diff --git a/src/lib/market-service/market-service-manager.ts b/src/lib/market-service/market-service-manager.ts
index 3847fbbd7a..c9b34f6204 100644
--- a/src/lib/market-service/market-service-manager.ts
+++ b/src/lib/market-service/market-service-manager.ts
@@ -50,6 +50,7 @@ export class MarketServiceManager {
     this.marketProviders = [
       // Order of this MarketProviders array constitutes the order of providers we will be checking first.
       // More reliable providers should be listed first.
+      new ZerionMarketService(),
       new CoinGeckoMarketService(),
       new CoinCapMarketService(),
       new PortalsMarketService(),
@@ -57,7 +58,6 @@ export class MarketServiceManager {
       // new YearnVaultMarketCapService({ yearnSdk }),
       // new YearnTokenMarketCapService({ yearnSdk }),
       new FoxyMarketService({ providerUrls, provider }),
-      new ZerionMarketService(),
       // TODO: Debank market provider
     ]
 
diff --git a/src/lib/market-service/zerion/zerion.ts b/src/lib/market-service/zerion/zerion.ts
index 4ce37f9b5c..bca240d557 100644
--- a/src/lib/market-service/zerion/zerion.ts
+++ b/src/lib/market-service/zerion/zerion.ts
@@ -9,11 +9,13 @@ import type {
 } from '@shapeshiftoss/types'
 import Axios from 'axios'
 import { setupCache } from 'axios-cache-interceptor'
+import type {
+  ZerionFungibles,
+  ZerionMarketData,
+} from 'components/TradeAssetSearch/hooks/useGetCustomTokenPriceQuery'
 
 import type { MarketService } from '../api'
 import { DEFAULT_CACHE_TTL_MS } from '../config'
-import type { ZerionFungibles, ZerionMarketData } from './types'
-import { zerionMarketDataToMarketData } from './utils'
 
 const axios = setupCache(Axios.create(), { ttl: DEFAULT_CACHE_TTL_MS, cacheTakeover: false })
 
@@ -41,10 +43,22 @@ export class ZerionMarketService implements MarketService {
   }
 
   async findByAssetId({ assetId }: MarketDataArgs): Promise<MarketData | null> {
+    const zerionMarketDataToMarketData = (zerionMarketData: ZerionMarketData): MarketData => {
+      return {
+        price: zerionMarketData.price.toString(),
+        marketCap: zerionMarketData.market_cap.toString(),
+        volume: '0', // Zerion doesn't provide volume data in this response
+        changePercent24Hr: zerionMarketData.changes.percent_1d,
+        supply: zerionMarketData.circulating_supply.toString(),
+        maxSupply: zerionMarketData.total_supply.toString(),
+      }
+    }
     try {
       const zerionMarketData = await this.getZerionTokenMarketData(assetId)
       if (!zerionMarketData) return null
-      return zerionMarketDataToMarketData(zerionMarketData)
+      const marketData = zerionMarketDataToMarketData(zerionMarketData)
+      if (!marketData) return null
+      return marketData
     } catch (e) {
       console.error(`MarketService(findByAssetId): error fetching market data: ${e}`)
       return null
```

File: src/lib/market-service/zerion/zerion.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  MarketDataArgs,
+  PriceHistoryArgs,
+} from '@shapeshiftoss/types'
+import Axios from 'axios'
+import { setupCache } from 'axios-cache-interceptor'
+
+import type { MarketService } from '../api'
+import { DEFAULT_CACHE_TTL_MS } from '../config'
+import type { ZerionFungibles, ZerionMarketData } from './types'
+import { zerionMarketDataToMarketData } from './utils'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts:16]: Doh! https://github.com/shapeshift/web/pull/7489/commits/1edee6c6784f4fbf2e6c5d0909feda596f96ff0e

File: src/lib/market-service/zerion/zerion.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts:31]: It'll stay last for now, but we should absolutely protect against it being moved! Excellent thought.

Addressed in https://github.com/shapeshift/web/pull/7489/commits/584a9e383855aa0792b5deeac1a1b8bc71701d58

Monkey patched to be the first provider, and it gracefully warns in the console and then falls back to the next provider:

Screenshot 2024-08-06 at 9 16 23 PM

File: src/lib/market-service/zerion/zerion.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  MarketData,
+  MarketDataArgs,
+  PriceHistoryArgs,
+} from '@shapeshiftoss/types'
+import Axios from 'axios'
+import { setupCache } from 'axios-cache-interceptor'
+
+import type { MarketService } from '../api'
+import { DEFAULT_CACHE_TTL_MS } from '../config'
+import type { ZerionFungibles, ZerionMarketData } from './types'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts:15]: Removed `useGetCustomTokenPriceQuery` (unused) in favour of the types file added in https://github.com/shapeshift/web/pull/7489/commits/1edee6c6784f4fbf2e6c5d0909feda596f96ff0e

File: src/lib/market-service/zerion/zerion.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    // TODO - unimplemented
+    return await Promise.resolve({})
+  }
+
+  async findPriceHistoryByAssetId({
+    assetId: _assetId,
+    timeframe: _timeframe,
+  }: PriceHistoryArgs): Promise<HistoryData[]> {
+    // TODO - unimplemented
+    return await Promise.resolve([])
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts]: Indeed! Addressed in https://github.com/shapeshift/web/pull/7489/commits/584a9e383855aa0792b5deeac1a1b8bc71701d58

File: src/lib/market-service/zerion/zerion.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+import type { MarketService } from '../api'
+import { DEFAULT_CACHE_TTL_MS } from '../config'
+import type { ZerionFungibles, ZerionMarketData } from './types'
+import { zerionMarketDataToMarketData } from './utils'
+
+const axios = setupCache(Axios.create(), { ttl: DEFAULT_CACHE_TTL_MS, cacheTakeover: false })
+
+export class ZerionMarketService implements MarketService {
+  baseUrl = 'https://zerion.shapeshift.com'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts]: Hahaha, that's fair - addressed in https://github.com/shapeshift/web/pull/7489/commits/536b7ad84e55fd26d456fc29e163f55466ca3659.

File: src/lib/market-service/zerion/zerion.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  getZerionTokenMarketData = async (assetId: AssetId): Promise<ZerionMarketData | undefined> => {
+    const { assetReference } = fromAssetId(assetId)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts:26]: Right you are ser - excluded non-token assets in https://github.com/shapeshift/web/pull/7489/commits/8b214fa0d7834d3f92bea24e97c98ecac03e6f40

File: src/lib/market-service/zerion/zerion.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    const url = `${this.baseUrl}/fungibles/${assetReference}`
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/market-service/zerion/zerion.ts:29]: As far as I can tell the [Zerion API](https://developers.zerion.io/reference/getfungiblebyid) doesn't support such filters, and does magic to have only one asset at an address.

Screenshot 2024-08-06 at 9 43 19 PM

Screenshot 2024-08-06 at 9 43 44 PM

[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/lib/market-service/zerion/zerion.ts:29]: That's `/fungibles/` allowing that https://developers.zerion.io/reference/listfungibles though on second thought, probably disregard me: that's super edge-case, and other market-data providers would handle those assets most likely🙏🏽

File: src/lib/market-service/zerion/types.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  circulating_supply: number
+  fully_diluted_valuation: number
+  market_cap: number
+  changes: {
+    percent_1d: number
+    percent_30d: number
+    percent_90d: number
+    pecent_365d: number
+  }
+}
```
Comment (gomesalexandre) [inline src/lib/market-service/zerion/types.ts:13]: preferably-blocking: `number | null` all fields
image

**Review by gomesalexandre — APPROVED** [review]
Retested locally with mainnet STETH (`0xae7ab96520de3a18e5e111b5eaab095312d7fe84`) monkey patched out of `generatedAssetdata.json`, LGTM!

One comment re: types to tackle at your convenience 🙏🏽

https://jam.dev/c/044c76fa-e499-453d-9720-53a076340482

File: src/lib/market-service/zerion/types.ts (hunk 11 above)
Comment (0xApotheosis) [reply src/lib/market-service/zerion/types.ts]: Good call - addressed in https://github.com/shapeshift/web/pull/7489/commits/f5f012c79bf83ad8adad054e5cef596ec90910be & https://github.com/shapeshift/web/pull/7489/commits/d6b682c34c479bce0832e70dbf662efa3079f68a

**Pull request merged**